facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+        // here, in case the entire chunk doesn't get uploaded successfully.
+        let body = Cursor::new(chunk.clone());
+
+        let http_response = ureq::put(&self.upload_session_uri)
+            .set("Content-Length", &format!("{}", chunk_len))
+            .set("Content-Range", &content_range)
+            // By default, ureq will wait forever to connect or read
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(10_000) // ten seconds
+            .send(body);
+
+        // On success we expect HTTP 308 Resume Incomplete and a Range: header,
+        // unless this is the last part and the server accepts the entire
+        // provided Content-Range, in which case it's HTTP 200, or 201 (?).
+        // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+        match http_response.status() {
+            200 | 201 => {
+                if !last_chunk {
+                    return Err(anyhow!(
+                        "received HTTP 200 or 201 response with chunks remaining"
+                    ));
+                }
+            }
+            308 => {
+                if !http_response.has("Range") {
+                    return Err(anyhow!(
+                        "No range header in response from GCS: {:?}",
+                        http_response.into_string()
+                    ));
+                }
+                let range_header = http_response.header("Range").unwrap();
+                // The range header is like "bytes=111-222", and represents a
+                // range in the overall object, not the current chunk.
+                let range = range_header
+                    .strip_prefix("bytes=")
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let mut values = range.splitn(2, '-');
+                let start = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let end = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+
+                if start != self.object_upload_position {
+                    return Err(anyhow!("unexpected Range header {}", range_header));
+                }
+
+                self.object_upload_position = end + 1;
+
+                // If we have a little content left over, we can't just make
+                // another request, because if there's too little of it, Google
+                // will reject it. Instead, put the portion of the chunk that we
+                // didn't manage to upload back into self.buffer so it can be
```

Comment:
I think this use case is a great match for `VecDeque`. You can `.send_bytes()` on an appropriately-sized slice of the VecDeque, then `pop_front` however many bytes actually got written.